### ISSUE TITLE
pdal pipeline filter_crop leads to SIGABRT

### ISSUE THREAD

**Issue by neteler** [PDAL/PDAL#3965]
**Describe the bug**

The attempt to clip this LAZ file:
https://www.opengeodata.nrw.de/produkte/geobasis/hm/3dm_l_las/3dm_l_las/3dm_32_376_5679_1_nw.laz

to a spatial subset leads to a 

```
$ pdal pipeline filter_crop.json --nostream
```

The filter file `filter_crop.json`:

```
[
    "3dm_32_376_5679_1_nw.laz",
    {
        "type":"filters.crop",
        "bounds":"([375875,376069],[5679235,5679510])"
    },
    {
        "type":"writers.las",
        "compression":"lazperf",
        "filename":"crop_3dm_32_376_5679_1_nw.laz"
    }
]
```

**Expected behavior**

Written LAZ file, clipped to the specified square.

**PDAL Fedora spec file**

My system is Fedora 37. This is the SPEC file used to build the official Fedora package:
https://src.fedoraproject.org/rpms/PDAL/blob/rawhide/f/PDAL.spec

**System/installation information:**

```
$ uname -a
Linux caddy 6.1.7-200.fc37.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 18 17:11:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ pdal --version
pdal 2.4.3 (git-version: Release)


gdb pdal
GNU gdb (GDB) Fedora Linux 12.1-7.fc37
[...]
This GDB was configured as "x86_64-redhat-linux-gnu".
[...]
Reading symbols from pdal...
[...]
(gdb) r pipeline filter_crop.json --nostream                                                                                                             
Starting program: /usr/bin/pdal pipeline filter_crop.json --nostream
/usr/include/c++/12/array:208: constexpr std::array<_Tp, _Nm>::value_type& std::array<_Tp, _Nm>::operator[](size_type) [with _Tp = long unsigned int; long unsigned int _Nm = 15; reference = long unsigned int&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.

Program received signal SIGABRT, Aborted.
__pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
Downloading 0.00 MB source file /usr/src/debug/glibc-2.36-9.fc37.x86_64/nptl/pthread_kill.c
44	      return INTERNAL_SYSCALL_ERROR_P (ret) ? INTERNAL_SYSCALL_ERRNO (ret) : 0;
(gdb) bt full
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
        tid = <optimized out>
        ret = 0
        pd = <optimized out>
        old_mask = {__val = {0}}
        ret = <optimized out>
#1  0x00007ffff72afec3 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
No locals.
#2  0x00007ffff725fa76 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
        ret = <optimized out>
#3  0x00007ffff72497fc in __GI_abort () at abort.c:79
        save_stage = 1

        act = {__sigaction_handler = {sa_handler = 0x20, sa_sigaction = 0x20}, sa_mask = {__val = {42949672969, 140737350497676, 208, 140737350562176, 
              140737350373376, 140737488341648, 9329128619081608960, 0, 18446744073709551456, 0, 9329128619081608960, 140737488341648, 
              9329128619081608960, 0, 18446744073709551456, 0}}, sa_flags = 1453374576, sa_restorer = 0x1b}
#4  0x00007ffff74d7e80 in std::__glibcxx_assert_fail (file=file@entry=0x7ffff7c8758c "/usr/include/c++/12/array", line=line@entry=208, 
    function=function@entry=0x7ffff7c97180 "constexpr std::array<_Tp, _Nm>::value_type& std::array<_Tp, _Nm>::operator[](size_type) [with _Tp = long unsigned int; long unsigned int _Nm = 15; reference = long unsigned int&; size_type = long unsi"..., 
    condition=condition@entry=0x7ffff7c69000 "__n < this->size()") at ../../../../../libstdc++-v3/src/c++11/debug.cc:60
No locals.
#5  0x00007ffff7bab197 in std::array<unsigned long, 15ul>::operator[] (this=<optimized out>, __n=<optimized out>) at /usr/include/c++/12/array:208
        __PRETTY_FUNCTION__ = <optimized out>
#6  std::array<unsigned long, 15ul>::operator[] (__n=18446744073709551615, this=0x555555642038) at /usr/include/c++/12/array:206
        __PRETTY_FUNCTION__ = <optimized out>
#7  pdal::las::Summary::addPoint (this=0x555555642008, x=<optimized out>, y=<optimized out>, z=<optimized out>, returnNumber=0)
    at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/io/private/las/Summary.cpp:60
No locals.
#8  0x00007ffff7ad91df in pdal::LasWriter::fillPointBuf (this=this@entry=0x555555641380, point=..., ostream=...)
    at /usr/include/c++/12/bits/unique_ptr.h:191
        has14PointFormat = <optimized out>
        returnNumber = <optimized out>
        numberOfReturns = <optimized out>
        converter = {__this = 0x555555641380}
        xOrig = <optimized out>
        yOrig = <optimized out>
        zOrig = <optimized out>
        x = <optimized out>
        y = <optimized out>
        z = <optimized out>
        scanChannel = <optimized out>
        scanDirectionFlag = <optimized out>
        edgeOfFlightLine = <optimized out>
        classification = <optimized out>
        userData = <optimized out>
        e = {f = -3.4851401e+33, d = 6.9533484896132662e-310, s8 = -89 '\247', s16 = -11097, s32 = -148122457, s64 = 140737340232871, u8 = 167 '\247', 
          u16 = 54439, u32 = 4146844839, u64 = 140737340232871}
        maxReturnCount = 5
#9  0x00007ffff7ad9bd3 in pdal::LasWriter::fillWriteBuf (this=0x555555641380, view=..., startId=<optimized out>, buf=...)
    at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/io/LasWriter.cpp:966
        idx = 498
        blocksize = 29411
        lastId = 29411
        ostream = {<pdal::Inserter> = {_vptr.Inserter = 0x7ffff7d73d30 <vtable for pdal::LeInserter+16>, 
            m_pbase = 0x5555aff3d820 "&\273=\002\036\325\331!7V", m_epptr = 0x5555b0031a60 "", m_pptr = 0x5555aff41a66 ""}, <No data fields>}
        point = {m_container = 0x555555e17290, m_layout = 0x555555611800, m_idx = 498, m_tmp = false}
#10 0x00007ffff7ad9e23 in pdal::LasWriter::writeView (this=0x555555641380, view=std::shared_ptr<pdal::PointView> (use count 7, weak count 0) = {...})
    at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/io/LasWriter.cpp:809
        filled = <optimized out>
        pointLen = 34
        viewRef = <optimized out>
        remaining = 720872
        idx = 0
#11 0x00007ffff7a78357 in pdal::FlexWriter::write (this=0x555555641380, view=std::shared_ptr<pdal::PointView> (use count 7, weak count 0) = {...})
    at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/pdal/FlexWriter.hpp:134
No locals.
#12 0x00007ffff7a70c4c in virtual thunk to pdal::Writer::run(std::shared_ptr<pdal::PointView>) ()
    at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/pdal/Writer.hpp:80
        pdal::s_info = {<pdal::PluginInfo> = {name = "writers.bpf", 
            description = "\"Binary Point Format\" (BPF) writer support. BPF is a simple \nDoD and research format that is used by some sensor and \nprocessing chains.", link = "http://pdal.io/stages/writers.bpf.html"}, extensions = std::vector of length 1, capacity 1 = {"bpf"}}
#13 0x00007ffff7bc1f39 in pdal::StageRunner::run (this=0x55559c16af50) at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/pdal/private/StageRunner.cpp:58
        keepSize = 720872
#14 0x00007ffff7b7414b in pdal::Stage::execute (this=0x555555641570, table=..., views=...) at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/pdal/Stage.cpp:287
        r = std::shared_ptr<pdal::StageRunner> (use count 2, weak count 0) = {get() = <optimized out>}
        __for_range = <optimized out>
        __for_begin = <optimized out>
        __for_end = <optimized out>
        outViews = std::set with 0 elements
        runners = <optimized out>
        srs = {m_wkt = "", m_horizontalWkt = ""}
        keeps = std::set with 1 element = {[0] = std::shared_ptr<pdal::PointView> (use count 7, weak count 0) = {get() = 0x555555e17290}}
#15 0x00007ffff7b74cba in pdal::Stage::execute (this=<optimized out>, table=...) at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/pdal/Stage.cpp:232
        si = {m_stage = 0x555555641570, m_id = 1}
        inViews = std::set with 1 element = {[0] = std::shared_ptr<pdal::PointView> (use count 7, weak count 0) = {get() = 0x555555e17290}}
        child = {m_stage = 0x555555641570, m_id = 0}
        stageInstanceId = <optimized out>
        stages = std::stack wrapping: std::deque with 0 elements
        pending = std::stack wrapping: std::deque with 0 elements
        children = std::map with 2 elements = {[{m_stage = 0x555555640f48, m_id = 2}] = {m_stage = 0x555555641570, m_id = 1}, [{
            m_stage = 0x555555640808, m_id = 3}] = {m_stage = 0x555555640f48, m_id = 2}}
        outViews = std::set with 1 element = {[0] = std::shared_ptr<pdal::PointView> (use count 7, weak count 0) = {get() = 0x555555e17290}}
        sets = std::map with 1 element = {[{m_stage = 0x555555641570, m_id = 1}] = std::set with 1 element = {
            [0] = std::shared_ptr<pdal::PointView> (use count 7, weak count 0) = {get() = 0x555555e17290}}}
#16 0x00007ffff7b4ce64 in pdal::PipelineManager::execute (this=0x55555562f4f8, mode=<optimized out>)
    at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/pdal/PipelineManager.cpp:273
        cnt = <optimized out>
        result = {m_mode = pdal::ExecMode::None, m_count = 0}
        s = <optimized out>
#17 0x00007ffff7b281ff in pdal::PipelineKernel::execute (this=0x55555562f4e0) at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/kernels/PipelineKernel.cpp:153
No locals.
#18 0x00007ffff7b42385 in pdal::Kernel::doExecution (this=<optimized out>, args=...) at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/pdal/Kernel.cpp:247
        status = 1
#19 0x00007ffff7b45f46 in pdal::Kernel::run (this=this@entry=0x55555562f4e0, cmdArgs=std::vector of length 2, capacity 2 = {...}, 
    log=std::shared_ptr<pdal::Log> (use count 9, weak count 0) = {...}) at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/pdal/Kernel.cpp:235
        args = {m_args = std::vector of length 14, capacity 16 = {std::unique_ptr<pdal::Arg> = {get() = 0x55555563c450}, std::unique_ptr<pdal::Arg> = {
              get() = 0x55555563c5e0}, std::unique_ptr<pdal::Arg> = {get() = 0x55555563c720}, std::unique_ptr<pdal::Arg> = {get() = 0x55555563c8b0}, 
            std::unique_ptr<pdal::Arg> = {get() = 0x55555563c9d0}, std::unique_ptr<pdal::Arg> = {get() = 0x55555563cbd0}, std::unique_ptr<pdal::Arg> = {
              get() = 0x55555563cd60}, std::unique_ptr<pdal::Arg> = {get() = 0x55555563cee0}, std::unique_ptr<pdal::Arg> = {get() = 0x55555563d0e0}, 
            std::unique_ptr<pdal::Arg> = {get() = 0x55555563d220}, std::unique_ptr<pdal::Arg> = {get() = 0x55555563d390}, std::unique_ptr<pdal::Arg> = {
              get() = 0x55555563d4f0}, std::unique_ptr<pdal::Arg> = {get() = 0x55555563d630}, std::unique_ptr<pdal::Arg> = {get() = 0x55555563d790}}, 
          m_shortargs = std::map with 2 elements = {["i"] = 0x55555563c9d0, ["s"] = 0x55555563d220}, m_longargs = std::map with 14 elements = {
            ["developer-debug"] = 0x55555563c450, ["dims"] = 0x55555563d790, ["driver"] = 0x55555563c720, ["help"] = 0x55555563c8b0, 
            ["input"] = 0x55555563c9d0, ["label"] = 0x55555563c5e0, ["metadata"] = 0x55555563d630, ["nostream"] = 0x55555563d4f0, 
            ["pipeline-serialization"] = 0x55555563cbd0, ["pointcloudschema"] = 0x55555563d0e0, ["progress"] = 0x55555563cee0, 
            ["stdin"] = 0x55555563d220, ["stream"] = 0x55555563d390, ["validate"] = 0x55555563cd60}}
        startup_status = 0
#20 0x000055555555e98e in App::execute (log=std::shared_ptr<pdal::Log> (use count 9, weak count 0) = {...}, 
    cmdArgs=std::vector of length 2, capacity 2 = {...}, this=0x7fffffffd5b0) at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/apps/pdal.cpp:371
        ret = 0
        name = "kernels.pipeline"
        kernel = 0x55555562f4e0
        args = {m_args = std::vector of length 11, capacity 16 = {std::unique_ptr<pdal::Arg> = {get() = 0x55555562df00}, std::unique_ptr<pdal::Arg> = {
              get() = 0x55555562e080}, std::unique_ptr<pdal::Arg> = {get() = 0x55555562e210}, std::unique_ptr<pdal::Arg> = {get() = 0x55555562e3b0}, 
            std::unique_ptr<pdal::Arg> = {get() = 0x55555562e4d0}, std::unique_ptr<pdal::Arg> = {get() = 0x55555562e6b0}, std::unique_ptr<pdal::Arg> = {
              get() = 0x55555562e7f0}, std::unique_ptr<pdal::Arg> = {get() = 0x55555562e930}, std::unique_ptr<pdal::Arg> = {get() = 0x55555562eb50}, 
            std::unique_ptr<pdal::Arg> = {get() = 0x55555562ed20}, std::unique_ptr<pdal::Arg> = {get() = 0x55555562ee20}}, 
          m_shortargs = std::map with 2 elements = {["h"] = 0x55555562e4d0, ["v"] = 0x55555562e210}, m_longargs = std::map with 11 elements = {
            ["command"] = 0x55555562df00, ["debug"] = 0x55555562e080, ["drivers"] = 0x55555562e3b0, ["help"] = 0x55555562e4d0, 
            ["list-commands"] = 0x55555562e6b0, ["log"] = 0x55555562eb50, ["logtiming"] = 0x55555562ed20, ["options"] = 0x55555562e930, 
            ["showjson"] = 0x55555562ee20, ["verbose"] = 0x55555562e210, ["version"] = 0x55555562e7f0}}
        args = <optimized out>
        e = <optimized out>
        ret = <optimized out>
        name = <optimized out>
        kernel = <optimized out>
#21 main (argc=<optimized out>, argv=<optimized out>) at /usr/src/debug/PDAL-2.4.3-1.fc37.x86_64/apps/pdal.cpp:315
        pdal = {m_out = @0x55555558c040, m_command = "pipeline", m_debug = false, m_logLevel = pdal::LogLevel::None, m_showDrivers = false, 
          m_help = false, m_showCommands = false, m_showVersion = false, m_showOptions = "", m_showJSON = false, m_log = "stderr", m_logtiming = false}
        cmdArgs = std::vector of length 2, capacity 2 = {"filter_crop.json", "--nostream"}

**Comment by abellgithub** [comment]
FYI, your file contains invalid return numbers.

Also, you must be building with some sort of debug assertion for array bounds for this to happen. Great to catch this issue, but potentially less good for performance.

**Comment by neteler** [comment]
> FYI, your file contains invalid return numbers.

Oh. This are official federal state data, guess I can't do much about it.
Thanks for having tested the file.

PDAL is my only tool to deal with the data :)

> Also, you must be building with some sort of debug assertion for array bounds for this to happen. Great to catch this issue, but potentially less good for performance.

Maybe @manisandro (co-packager) has an idea here?

**Comment by hobu** [comment]
As an aside, this scenario is a good case for using COPC, especially if you have to do more than one crop of the data.

```
$ time pdal pipeline las.json 
51.78s user 0.29s system 99% cpu 52.165 total
```

```
$ time pdal pipeline copc.json
5.05s user 0.07s system 203% cpu 2.511 total
```

**Comment by hobu** [comment]
@abellgithub addressed the specific crash with #3966 which will be released in 2.5.1 which is expected next week sometime.

**Comment by neteler** [comment]
Thanks so much for the quick fix with broken input data!
